api/applications_api.py

```diff
@@ -0,0 +1,188 @@
+from datetime import datetime
+from dateutil.parser import parse
+from django.shortcuts import get_object_or_404
+from django.utils import timezone
+from rest_framework import viewsets, serializers
+from rest_framework.response import Response
+from applications.models import (
+    Application,
+    Organisation,
+    ApplicationEvent,
+    Recurrence,
+    Address,
+    Person,
+)
+
+
+MINIMUM_TIME = timezone.datetime(
+    1970, 1, 1, 0, 0, 0, 3, timezone.get_default_timezone()
+)
+
+MAXIMUM_TIME = timezone.datetime(
+    2099, 1, 1, 0, 0, 0, 3, timezone.get_default_timezone()
+)
+
+
+class PersonSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Person
+
+        fields = "__all__"
+
+
+class PersonViewSet(viewsets.ModelViewSet):
+
+    queryset = Person.objects.all()
+
+    serializer_class = PersonSerializer
+
+    def perform_create(self, serializer):
+        serializer.save()
+
+
+class AddressSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Address
+
+        fields = "__all__"
+
+
+class AddressViewSet(viewsets.ModelViewSet):
+
+    queryset = Address.objects.all()
+
+    serializer_class = AddressSerializer
+
+    def perform_create(self, serializer):
+        serializer.save()
+
+
+class OrganisationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Organisation
+
+        fields = ["id", "name", "identifier", "year_established"]
+
+
+class OrganisationViewSet(viewsets.ModelViewSet):
+
+    queryset = Organisation.objects.all()
+
+    serializer_class = OrganisationSerializer
+
+    def perform_create(self, serializer):
+        serializer.save()
+
+
+class ApplicationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Application
+        fields = [
+            "id",
+            "description",
+            "reservation_purpose",
+            "organisation",
+            "contact_person",
+            "user",
+            "organisation",
+        ]
+
+    def validate(self, data):
+        return data
+
+
+class ApplicationViewSet(viewsets.ModelViewSet):
+
+    queryset = Application.objects.all()
+
+    serializer_class = ApplicationSerializer
+
+    def perform_create(self, serializer):
+        serializer.save(user=self.request.user)
+
+
+class RecurrenceReadSerializer(serializers.ModelSerializer):
+    def to_representation(self, data):
+        recurrence = {}
+        for field in self.fields:
+            recurrence[field] = getattr(data, field)
+
+        recurrence["recurrence"] = data.recurrence.between(
+            self.get_start_date(),
+            self.get_end_date(),
+        )
+        return recurrence
+
+    def get_start_date(self) -> datetime:
```

Comment:
So this and get end date have some problems. The start (context) is set in the ApplicationEventViewSet#get_request_params

It works perfectly when i call /applications/ to fetch all the applications but when i do applications/id to fetch single application. I get like 4 extra times that end up in 

return MINIMUM_TIME

branch even start is set correctly in ApplicationEventViewSet. The first call to get_start_date has start date in context, but for some reason there are some extra calls to this method where the context is not set. Even with those extra calls the api works just fine and returns recurrences only between given start and end dates. Can't figure out where those extra calls come. 

And as i said there are extra call only when fetching single one, not when fetching all. And in both cases api returns correctly.